File: ring-android/app/src/main/java/cx/ring/settings/PluginDetails.java
package cx.ring.settings;

import android.graphics.drawable.Drawable;

import java.io.File;
import java.util.List;
import java.util.Map;

import cx.ring.daemon.Ringservice;
import cx.ring.utils.Log;

/**
 * Class that contains PluginDetails like name, rootPath
 */
public class PluginDetails {
    public static final String TAG = PluginDetails.class.getSimpleName();
    private String name;
    private String rootPath;
    private boolean enabled;

    public PluginDetails(String name, String rootPath, boolean enabled) {
        this.name = name;
        this.rootPath = rootPath;
        this.enabled = enabled;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getRootPath() {
        return rootPath;
    }

    /**
     * @return String: absolute path to the so file
     */
    public String getSoPath() {
        return rootPath + "/lib" + name + ".so";
    }

    public void setRootPath(String rootPath) {
        this.rootPath = rootPath;
    }

    /**
     * Returns the plugin activation status by the user
     * @return boolean
     */
    public boolean isEnabled() {
        return enabled;
    }

    public void setEnabled(boolean enabled) {
        this.enabled = enabled;
    }

    public Drawable getPluginIcon() {
        String iconPath = Ringservice.getPluginIconPath(getSoPath());
        Drawable icon = null;

        File file = new File(iconPath);

        Log.i(TAG, "Plugin icon path: " + iconPath);

        if(file.exists()) {
            icon = Drawable.createFromPath(iconPath);
        }

        return icon;
    }

    public List<Map<String,String>> getPluginPreferences() {
        return Ringservice.getPluginPreferences(getSoPath()).toNative();
    }

    public Map<String, String> getPluginPreferencesValues() {
        return Ringservice.getPluginPreferencesValuesMap(getSoPath()).toNative();
    }

    public boolean setPluginPreference(String key, String value) {
        return Ringservice.setPluginPreference(getSoPath(), key, value);
    }

}

File: jami-android/app/src/main/java/cx/ring/settings/pluginssettings/PluginDetails.kt
/*
 *  Copyright (C) 2004-2023 Savoir-faire Linux Inc.
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package cx.ring.settings.pluginssettings

import android.graphics.drawable.Drawable
import android.util.Log
import cx.ring.fragments.CallFragment
import cx.ring.utils.ConversationPath
import net.jami.daemon.JamiService
import java.io.File
import java.util.*

/**
 * Class that contains PluginDetails like name, rootPath
 */
class PluginDetails(val name: String, val rootPath: String, var isEnabled: Boolean, var handlerId: String?= null, var accountId: String? = "") {
    private val details: Map<String, String> = pluginDetails
    var icon: Drawable? = null
        private set
    var isRunning: Boolean = false

    fun setIcon() {
        var iconPath = details["iconPath"]
        if (iconPath != null) {
            if (iconPath.endsWith("svg"))
                iconPath = iconPath.replace(".svg", ".png")
            val file = File(iconPath)
            if (file.exists()) {
                icon = Drawable.createFromPath(iconPath)
            }
        }
    }

    private val pluginDetails: Map<String, String>
        get() = JamiService.getPluginDetails(rootPath).toNative()
    val pluginPreferences: List<Map<String, String>>
        get() = JamiService.getPluginPreferences(rootPath, accountId).toNative()
    val pluginPreferencesValues: Map<String, String>
        get() = JamiService.getPluginPreferencesValues(rootPath, accountId)

    fun setPluginPreference(key: String, value: String): Boolean {
        return JamiService.setPluginPreference(rootPath, accountId, key, value)
    }

    companion object {
        val TAG = PluginDetails::class.simpleName!!
    }

    init {
        setIcon()
    }
}